src/Agent.js

```diff
@@ -0,0 +1,187 @@
+import request from 'superagent';
+import prefix from 'superagent-prefix';
+
+
+export default class Agent {
+
+	constructor(baseUrl) {
+		this.prefix = prefix(baseUrl);
+	}
+
+	get(uri, auth, query) {
+		return this.request({uri, auth, method: 'get', query: query});
+	}
+
+	head(uri, auth) {
+		return this.request({uri, auth, method: 'head'});
+	}
+
+	post(uri, data, auth) {
+		return this.request({uri, data, auth, method: 'post'});
+	}
+
+	put(uri, data, auth) {
+		return this.request({uri, data, auth, method: 'put'});
+	}
+
+	delete(uri, data, auth) {
+		return this.request({uri, data, auth, method: 'delete'});
+	}
+
+
+	/**
+	 *
+	 * @param {String} uri           The URI to request
+	 * @param {String} method        The method used to request the URI, should be in uppercase.
+	 * @param {String} data          Arbitrary data to send as the body.
+	 * @param {Object} auth          Authorization
+	 * @param {String} query         Query parameters
+	 * @param {Object} form          Form fields
+	 * @param {Object} files         array of file names and file content
+	 * @return {Promise} A promise. fulfilled with {body, statusCode}, rejected with { statusCode, errorDescription, error, body }
+	 */
+	request({uri, method, data = undefined, auth, query = undefined, form = undefined, files = undefined}) {
+		const requestFiles = this._sanitizeFiles(files);
+		return this._request({uri, method, data, auth, query, form, files: requestFiles});
+	}
+
+	/**
+	 *
+	 * @param {String} uri           The URI to request
+	 * @param {String} method        The method used to request the URI, should be in uppercase.
+	 * @param {String} data          Arbitrary data to send as the body.
+	 * @param {Object} auth          Authorization
+	 * @param {String} query         Query parameters
+	 * @param {Object} form          Form fields
+	 * @param {Object} files         array of file names and file content
+	 * @return {Promise} A promise. fulfilled with {body, statusCode}, rejected with { statusCode, errorDescription, error, body }
+	 */
+	_request({uri, method, data, auth, query, form, files}) {
+		const req = this._buildRequest({uri, method, data, auth, query, form, files});
+		return this._promiseResponse(req);
+	}
+
+	/**
+	 * Promises to send the request and retreive the response.
+	 * @param {Request} req The request to send
+	 * @returns {Promise}   The promise to send the request and retrieve the response.
+	 * @private
+	 */
+	_promiseResponse(req) {
+		return new Promise((fulfill, reject) => this._sendRequest(req, fulfill, reject));
+	}
+
+	/**
+	 * Sends the given request, calling the fulfill or reject methods for success/failure.
+	 * @param {object} request   The request to send
+	 * @param {function} fulfill    Called on success with the response
+	 * @param {function} reject     Called on failure with the failure reason.
+	 * @private
+	 * @returns {undefined} Nothing
+	 */
+	_sendRequest(request, fulfill, reject) {
+		request.end((error, res) => {
```

Comment:
I recently learned that `superagent` supports a `then` method, so you could promise up this method if you wanted.